docs/variables.md

```diff
@@ -0,0 +1,73 @@
+# Variables #
+
+There are different ways to declare a variable `a` of type `α` in Lean.
+
+## () vs {}
+Consider the following two ways to define the projection of a length 2 vector onto its first component.
+
+```lean
+import data.vector
+
+variable {α : Type}
+
+def p₁: vector α 2 → α
+| v := vector.head v
+#check p₁ -- p₁ : vector ?M_1 2 → ?M_1
+#check @p₁ -- p₁ : Π {α : Type}, vector α 2 → α
+
+def v : vector nat 2 := 1 :: 2 :: vector.nil
+#eval p₁ v -- 1
+
+variable (β : Type) -- or equivalently variable β : Type
+
+def π₁: vector β 2 → β
+| v := vector.head v
+#check π₁ -- π₁ : Π (β : Type), vector β 2 → β
+#check @π₁ -- π₁ : Π (β : Type), vector β 2 → β
+
+#eval π₁ nat v -- 1
+```
+
+Curly braces tell Lean that the following variable may be inferred. Whenever one calls `p₁ v` with `v : vector nat 2` it is clear that `α` in the definition of `p₁` is `nat`. Use the `@` operator to check for those "hidden" variables. Hence `#check @p₁` and `#check @π₁` yield almost the same result. If one wants to make an implicit variable like `{α : Type}` explicit later on in the same file one can use `variable (α)`.
+
+The `variables` keyword is shorthand notation and can be circumvented as follows:
+
+```lean
+def p₁ {α : Type}: vector α 2 → α := sorry
+#check p₁ -- p₁ : vector ?M_1 2 → ?M_1
+#check @p₁ -- p₁ : Π {α : Type}, vector α 2 → α
+
+def π₁ (β : Type): vector β 2 → β := sorry
+#check π₁ -- π₁ : Π (β : Type), vector β 2 → β
+#check @π₁ -- π₁ : Π (β : Type), vector β 2 → β
+```
+
+Nevertheless using `variables` may improve code clarity. Note that `variables` are somewhat "smart" in that definitions which do not depend on a previously introduced variable do not carry them in their signature: In the first example `π₁` does not depend on `α` at all.
+
+## []
+Square brackets refer to the [type class mechanism](https://leanprover.github.io/theorem_proving_in_lean/type_classes.html)
+
+## Pattern matching
```

Comment:
Thanks for your PR. I'm not sure if this section belongs in this file, or maybe should move to another/new file. Did you have a particular reason to put it in this file?